lib/src/Pos/Web/Server.hs

```diff
@@ -85,12 +87,36 @@ serveImpl
     => m Application
     -> String
     -> Word16
+    -- ^ if the port is 0, bind to a random port
     -> Maybe TlsParams
+    -- ^ if isJust, call it with the port after binding
     -> Maybe Settings
+    -> Maybe (Word16 -> IO ())
+    -- ^ if isJust, use https, isNothing, use raw http
     -> m ()
-serveImpl app host port mWalletTLSParams mSettings =
-    liftIO . maybe runSettings runTLS mTlsConfig mySettings =<< app
+serveImpl app host port mWalletTLSParams mSettings mPortCallback = do
+    app' <- app
+    let
+      acquire :: IO (Word16, Socket)
+      acquire = liftIO $ if port == 0
+        then do
+          (port', socket) <- bindRandomPortTCP (getHost mySettings)
+          pure (fromIntegral port', socket)
+        else do
+          socket <- bindPortTCP (fromIntegral port) (getHost mySettings)
+          pure (port, socket)
+      release :: (Word16, Socket) -> IO ()
+      release (_, socket) = liftIO $ close socket
+      action :: (Word16, Socket) -> IO ()
+      action (port', socket) = do
+        -- TODO: requires warp 3.2.17 setSocketCloseOnExec socket
+        launchServer app' (port', socket)
+    liftIO $ bracket acquire release action
   where
+    launchServer :: Application -> (Word16, Socket) -> IO ()
+    launchServer app'' (port', socket) = do
+      fromMaybe (const (pure ())) mPortCallback port'
+      maybe runSettingsSocket runTLSSocket mTlsConfig mySettings socket app''
```

Comment:
the project used 4 space indentation fwiw. consider using [EditorConfig](http://editorconfig.org/) to have it automatically do this for you :)

Reply:
i had changed my editor to use 4 on all haskell files, but the 2nd machine isnt updated yet